### PR TITLE
Use basic LLVM optimizations (PassManager) in JIT mode

### PR DESCRIPTION
**Pull request by EgorBo**
Contributes to https://github.com/mono/mono/issues/16235 

I just copied pieces from `opt.cpp` and old mono (old mono llvm 3.6 jit [used to support it](https://github.com/mono/mono/blob/1ec822976b92527fb57ef6d3ba6ccda2609d7db1/mono/mini/llvm-jit.cpp#L820-L838))
LLVM has tons of optimizations but it looks like not so much of them actually change anything for our IR yet, it can be checked this way: generate IR for different methods and feed `opt -O2 -print-before-all -print-after-all` with it. It should generate diffs for each transformation pass, e.g see: https://godbolt.org/z/QUszzv (`** IR Dump Before %pass%, ** IR Dump After  %pass%`...)

So this PR is just a minimal base (better than nothing, see `const char *opts =`)

### CODE REVIEW

File: scripts/ci/pipeline-netcore-runtime.yml [hunk 1]
```diff
@@ -157,7 +157,8 @@ stages:
 
         - bash: |
             cd netcore
-            ./build.sh
+            ./build.sh --llvm
+          timeoutInMinutes: 150
```
Comment (EgorBo) [inline scripts/ci/pipeline-netcore-runtime.yml]: Just to test LLVM JIT mode on all corefx tests in this PR on CI

File: netcore/Makefile [hunk 2]
```diff
@@ -197,7 +197,7 @@ run-tests-corefx-%: prepare update-tests-corefx
 	@cp corefx/restore/corefx-restore.dll corefx/tests/extracted/$*/corefx-restore.dll
 	@sed -i -e 's/5.0.0\"/$(NETCOREAPP_VERSION)\"/g' corefx/tests/extracted/$*/*.runtimeconfig.json
 	cd corefx/tests/extracted/$* && \
-	COMPlus_DebugWriteToStdErr=1 $(CURDIR)/./dotnet exec --runtimeconfig $*.runtimeconfig.json --fx-version "$(NETCOREAPP_VERSION)" xunit.console.dll $*.dll \
+	MONO_ENV_OPTIONS="--llvm" COMPlus_DebugWriteToStdErr=1 $(CURDIR)/./dotnet exec --runtimeconfig $*.runtimeconfig.json --fx-version "$(NETCOREAPP_VERSION)" xunit.console.dll $*.dll \
 		-html ../../TestResult-$*.html -xml ../../TestResult-$*-netcore-xunit.xml \
```
Comment (vargaz) [inline netcore/Makefile]: Unrelated change ?

Comment (EgorBo) [reply netcore/Makefile]: More like "temporarily" to run all corefx tests in this mode (will remove once CI finishes)

**Comment by EgorBo** [comment]
Benchmark results for https://github.com/aras-p/ToyPathTracer

image

File: mono/mini/llvm-jit.cpp [hunk 3]
```diff
@@ -247,24 +249,82 @@ init_mono_llvm_jit ()
...
+		//initializeInterleavedAccessPass(registry);
+		//initializeEntryExitInstrumenterPass(registry);
+		//initializePostInlineEntryExitInstrumenterPass(registry);
+		//initializeUnreachableBlockElimLegacyPassPass(registry);
+		//initializeExpandReductionsPass(registry);
+		//initializeWriteBitcodePassPass(registry);
+
+		// FIXME: find optimal mono specific order of passes, see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
+		const char *opts = "opt -simplifycfg -sroa -instcombine -early-cse-memssa -instcombine -gvn";
+		char **args = g_strsplit (opts, " ", -1);
```
Comment (vargaz) [inline mono/mini/llvm-jit.cpp]: What is "opt ?

File: mono/mini/llvm-jit.cpp [hunk 4]
```diff
@@ -247,24 +249,82 @@ init_mono_llvm_jit ()
...
+		  modules(),
+		  fpm (unwrap (mp)) {
+		initPassManager ();
+	}
+
+	void initPassManager () {
+		// Initialize passes (copied from https://github.com/mono/llvm/blob/release_60/tools/opt/opt.cpp#L379)
+		PassRegistry &registry = *PassRegistry::getPassRegistry();
+		initializeCore(registry);
+		//initializeCoroutines(registry);
```
Comment (vargaz) [inline mono/mini/llvm-jit.cpp]: Should probably delete the commented out lines.

File: mono/mini/llvm-jit.cpp (hunk 3 above)
Comment (EgorBo) [reply mono/mini/llvm-jit.cpp]: it ignores the first argument for some reason so it's fake

Comment (EgorBo) [reply mono/mini/llvm-jit.cpp]: Actually a leading space helps

**Comment by vargaz** [comment]
@monojenkins build failed
